src/devices/directv.c

```diff
@@ -0,0 +1,532 @@
+        }
+        return 0;
+    }
+    // We've received an acceptable decoded dtv value here.
+
+    // Quick and dirty, get bitbuffer row data into a pretty string.
+    uint8_t dtv_strbuf[35] = {0};  // probably only needs 25 bytes ({xx}twenty-nibbles-shown\0)
+    sprintf(dtv_strbuf, "{%2d}", DTV_BITLEN_MAX);
+    for (unsigned col = 0; col < (DTV_BITLEN_MAX + 7) / 8; ++col) {
+        sprintf(temp_strbuf, "%02x", dtv[col]);
+        strcat( dtv_strbuf, temp_strbuf );
+    }
+    if (decoder->verbose > 2) {
+        fprintf(stderr, "directv: decoded data value is: %s.\n", dtv_strbuf);
+    }
+
+    // Break apart dtv data into fields and validate
+    
+    // First byte should be 0x10 (magic number?)
+    if (dtv[0] != 0x10) {
+        if (decoder->verbose > 1) {
+            fprintf(stderr, "directv: Bad magic number: 0x%02x should be 0x10.\n", dtv[0]);
+        }
+        return 0;
+    }
+
+    // Validate Checksum
+    if ((dtv[0] >> 4 + dtv[0] & 0x0F + dtv[1] >> 4 + dtv[1] & 0x0F + dtv[2] >> 4 + dtv[2] & 0x0F + 
+        dtv[3] >> 4 + dtv[3] & 0x0F + dtv[4] >> 4) & 0x0F != dtv[4] & 0x0F)
+    {
+        if (decoder->verbose > 1) {
+            fprintf(stderr, "directv: Checksum failed.\n");
+        }
+        return 0;
+    }
+    
+    // Get Device ID 
+    unsigned dtv_device_id;
+    dtv_device_id = dtv[1] << 12 | dtv[2] << 4 | dtv[3] >> 4;
+    if (dtv_device_id > 999999) {
+        if (decoder->verbose > 1) {
+            fprintf(stderr, "directv: Bad Device ID: %d.\n", dtv_device_id);
+        }
+        return 0;
+    }
+
+    // Get Button ID
+    uint8_t dtv_button_id;
+    dtv_button_id = dtv[3] << 4 | dtv[4] >> 4;
+    // Assume all byte values are valid.
+
+    data = data_make(
+            "model",         "",            DATA_STRING, "DirecTV-RC66RX",
+            "device_id",     "",            DATA_FORMAT, "%06d", DATA_INT, dtv_device_id,
+            "button_id",     "",            DATA_FORMAT, "%d", DATA_INT, dtv_button_id,
+            "button_name",   "",            DATA_FORMAT, "[%s]", DATA_STRING, get_dtv_button_label(dtv_button_id),
+            "event",         "",            DATA_STRING, sync_type==0?"REPEAT":"INITIAL",
+            "mic",           "",            DATA_STRING, "CHECKSUM",
+            NULL);
+
+    decoder_output_data(decoder, data);
+
+    return 1;
+}
+
+static char *output_fields[] = {
+    "model",
+    "device_id",
+    "button_id",
+    "button_name",
+    "event",
+    "mic",
+    NULL
+};
+
+r_device directv = {
+        .name        = "DirecTV RC66RX Remote Control",
+        .modulation  = FSK_PULSE_PCM,
+        .short_width = 600,  // 150 samples @250k
+        .long_width  = 600,  // 150 samples @250k
+        .gap_limit   = 30000, // gap is typically around 27,600μs, so long that rtl_433 resets 
+                              // signal decoder before recognizing row repeats in signal
+        .reset_limit = 50000, // maximum gap size before End Of Row [μs]
```

Comment:
The maximum signal length heuristics messes up the row splitting here. One could think of using some kind of buffering here to get the actual rows. But that would mess with the realtime aspects of the remote. The only good idea I have here is to perform a duplicate check and suppress the duplicate events. That way we get a behavior that is usable and reliable for a consumer.

Reply:
My thought on how to implement this:

- Provide a function that will allow for generation of a hash value given any arbitrary text.
- Decoders can use this to create a hash value of whatever fields in their decode would be considered part of the same decoding if it were seen in the subsequent decoding.
- Allow decoders to put this hash value into their output data structure
- Provide this hash value somewhere accessible in the device or bitbuffer for the decoder to access.

In implementation, I'm thinking something like:

```
static int template_callback(r_device *decoder, bitbuffer_t *bitbuffer)
{

    [...]

    // Each decoder, if it wants, can add something like these two lines:
    sprintf(decode_identifying_data, "[New Template|%d|%d]", sensor_id, value);
    decode_hash = make_hash_value(decode_identifying_data);
    
    // Decoder has an option, if desired, to act on the repetition itself
    prev_decode_hash = bitbuffer->prev_decode_hash;
    if (decode_hash == prev_decode_hash) {
        do something
    }

    // But, it should just add the new hash here:
    data = data_make(
            "model", "", DATA_STRING, "New Template",
            "id",    "", DATA_INT,    sensor_id,
            "data",  "", DATA_INT,    value,
            "mic",   "", DATA_STRING, "CHECKSUM", // CRC, CHECKSUM, or PARITY
            "decode_hash", "", DATA_HASH, decode_hash,   // provide HASH here
            NULL);

    [...]
```

The main program can just choose to show all decodings as normal, but maybe 
there's a new flag to rtl_433 that will have it suppress all repetitions (just see
if decode hash is same as previous hash), or maybe count all the the repetitions
it sees and output them all as one second event with a field to denote this:

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2019-05-23 21:56:26
model     : DirecTV-RC66RX                         device_id : 123456
button_id : 0x25         button_name: [SELECT]     event     : INITIAL
mic       : CHECKSUM
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2019-05-23 21:56:28
model     : DirecTV-RC66RX                         device_id : 123456
button_id : 0x25         button_name: [SELECT]     event     : REPEAT
mic       : CHECKSUM     repetition_count: 15
```